### PR TITLE
SMU: Cumulative changes from Ian and Pat

### PR DESCRIPTION
**Pull request by pfrstg**
main changes
* Topology sensing now implemented in pipeline (expect an increase in run time)
* Format of SQLite database is compressed (new snappy dependency)
* Revision of the categorization and special field filtering / adjustment rules

### CODE REVIEW

**Review by kjk33 — CHANGES_REQUESTED** [review]

File: smu/geometry/utilities.py [hunk 1]
```diff
@@ -293,3 +323,116 @@ def is_single_fragment(bond_topology):
...
+               dataset_pb2.BondTopology.ATOM_O,
+               dataset_pb2.BondTopology.ATOM_F,
+               dataset_pb2.BondTopology.ATOM_H]:
+    return smu_utils_lib.ATOM_TYPE_TO_MAX_BONDS[atype]
+
+  if atype == dataset_pb2.BondTopology.ATOM_N:
+    return 4
+
+  if atype == dataset_pb2.BondTopology.ATOM_ONEG:
+    return 2
```
[the remark this answers is not in the document]
Comment (pfrstg) [reply smu/geometry/utilities.py:363]: Added a raise -- should never happen

File: smu/geometry/utilities.py [hunk 2]
```diff
@@ -293,3 +323,116 @@ def is_single_fragment(bond_topology):
...
+        bt.atoms[i] = dataset_pb2.BondTopology.ATOM_O
+    elif atom == dataset_pb2.BondTopology.ATOM_N:
+      if nbonds[i] == 4:
+        bt.atoms[i] = dataset_pb2.BondTopology.ATOM_NPOS
+        charges[i] = 1
+    elif atom == dataset_pb2.BondTopology.ATOM_NPOS:
+      if nbonds[i] == 3:
+        bt.atoms[i] = dataset_pb2.BondTopology.ATOM_N
+
+  return positive_placed == negative_placed
```
[the remark this answers is not in the document]
Comment (pfrstg) [reply smu/geometry/utilities.py:405]: This function is unused. I am deleting it.

File: smu/geometry/topology_from_geom.py [hunk 3]
```diff
@@ -139,45 +146,116 @@ def bond_topologies_from_geom(
...
 
+  score_sum = np.sum([bt.score for bt in result.bond_topology])
+  for bt in result.bond_topology:
+    bt.topology_score = np.log(bt.score / score_sum)
+    bt.ClearField("score")
+
   return result
 
+def geometry_score(bt: dataset_pb2.BondTopology,
+                   distances: np.array,
```
[the remark this answers is not in the document]
Comment (pfrstg) [reply smu/geometry/topology_from_geom.py:226]: fixed

File: smu/geometry/topology_from_geom.py [hunk 4]
```diff
@@ -139,45 +146,116 @@ def bond_topologies_from_geom(
   # with the score for each bond type.
 
   bonds_to_scores: Dict[Tuple[int, int], np.array] = {}
```
[the remark this answers is not in the document]
Comment (pfrstg) [reply smu/geometry/topology_from_geom.py:148]: fixed

File: smu/geometry/topology_from_geom.py [hunk 5]
```diff
@@ -139,45 +146,116 @@ def bond_topologies_from_geom(
...
     return result
 
+  # Need to know when the starting smiles has been recovered.
+  rdkit_mol = smu_utils_lib.bond_topology_to_molecule(bond_topology)
+  starting_smiles = smu_utils_lib.compute_smiles_for_molecule(
+    rdkit_mol, include_hs=True)
+  initial_ring_atom_count = utilities.ring_atom_count_mol(rdkit_mol)
+
+  # Avoid finding duplicates.
+  all_found_smiles: Set[string] = set()
```
[the remark this answers is not in the document]
Comment (pfrstg) [reply smu/geometry/topology_from_geom.py:173]: fixed

**Review by kjk33 — APPROVED** [review]

File: smu/parser/smu_utils_lib.py [hunk 6]
```diff
@@ -773,20 +782,16 @@ def _conformer_source(conf):
...
 MERGE_CONFLICT_FIELDS = [
     'conformer_id',
-    'error_nstat1_1',
-    'error_nstatc_1',
-    'error_nstatt_1',
-    'error_frequencies_1',
+    'error_nstat1',
+    'error_nstatc',
+    'error_nstatv',
+    'error_nstatt'
```
Comment (kjk33) [inline smu/parser/smu_utils_lib.py:788]: Missing comma at end of string. This leads to an implicit concat between this and the next line. Please review and fix.

Comment (pfrstg) [reply smu/parser/smu_utils_lib.py]: Good catch, fixed with 899e99b

**Pull request merged**